hazelcast-client-new/src/main/java/com/hazelcast/client/proxy/ClientMapProxy.java

```diff
@@ -1233,15 +1234,41 @@ public boolean isEmpty() {
 
     @Override
     public void putAll(Map<? extends K, ? extends V> m) {
-        Map<Data, Data> map = new HashMap<Data, Data>();
+        ClientPartitionService partitionService = getContext().getPartitionService();
+        Map<Integer, Map<Data, Data>> entryMap = new HashMap<Integer, Map<Data, Data>>(partitionService.getPartitionCount());
+        
         for (Entry<? extends K, ? extends V> entry : m.entrySet()) {
+            checkNotNull(entry.getKey(), NULL_KEY_IS_NOT_ALLOWED);
+            checkNotNull(entry.getValue(), NULL_VALUE_IS_NOT_ALLOWED);
+            
             final Data keyData = toData(entry.getKey());
             invalidateNearCache(keyData);
-            map.put(keyData, toData(entry.getValue()));
+            
+            int partitionId = partitionService.getPartitionId(keyData);
+            if (!entryMap.containsKey(partitionId)) {
+                entryMap.put(partitionId, new HashMap<Data, Data>());
+            }
+            
+            entryMap.get(partitionId).put(keyData, toData(entry.getValue()));
+        }
+        
+        List<Future<?>> futures = new ArrayList<Future<?>>(entryMap.size());
+        for (final Map.Entry<Integer, Map<Data, Data>> entry : entryMap.entrySet()) {
+            final Integer partitionId = entry.getKey();
+            //If there is only one entry, consider how we can use MapPutRequest
+            //without having to get back the return value.
+            ClientMessage request = MapPutAllCodec.encodeRequest(name, entry.getValue());
+            futures.add(new ClientInvocation(getClient(), request, partitionId).invoke());
         }
 
-        ClientMessage request = MapPutAllCodec.encodeRequest(name, map);
-        invoke(request);
+        try {
+            for (Future<?> future : futures) {
+                future.get();
```

Comment:
That's the intent - putAll does not specify a timeout, unlike the overrides for put.